### PR TITLE
chore: added validations and optional chaining in customerIo to avoid bugsnag Issue

### PR DESCRIPTION
**Pull request by anantjain45823**
## Description of the change

> We are seeing too many bugsnag alerts in respect of undefined or NULL fields.
To avoid these we are thinking of introducing additional checks for all integrations starting with customerIo destination.
In this PR I have included some validations and optional chaining in customerIo destination 

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/customerio/util.js [hunk 1]
```diff
@@ -295,6 +297,18 @@ const defaultResponseBuilder = (message, evName, userId, evType, destination, me
   return { rawPayload, endpoint, requestConfig };
 };
 
+const validateConfigFields = destination => {
+  const { Config } = destination;
+  if (!isDefinedAndNotNull(Config)) {
+    throw new ConfigurationError(" No Configs Found");
```
Comment (mihir-4116) [inline src/v0/destinations/customerio/util.js:303]: why is this ConfigurationError?
when this case will happen?
is there any other places where we are handling this?

File: src/v0/destinations/customerio/util.js [hunk 2]
```diff
@@ -295,6 +297,18 @@ const defaultResponseBuilder = (message, evName, userId, evType, destination, me
   return { rawPayload, endpoint, requestConfig };
 };
 
+const validateConfigFields = destination => {
```
Comment (mihir-4116) [inline src/v0/destinations/customerio/util.js:300]: add function description

File: src/v0/destinations/customerio/util.js [hunk 3]
```diff
@@ -236,13 +237,14 @@ const defaultResponseBuilder = (message, evName, userId, evType, destination, me
...
   const isDeviceRelatedEvent = isdeviceRelatedEventName(evName, destination);
   if (isDeviceRelatedEvent && userId && token) {
+    const timestamp = message.timestamp || message.originalTimestamp;
     const devProps = {
       ...message.properties,
       id: token,
-      last_used: Math.floor(new Date(message.originalTimestamp).getTime() / 1000),
+      last_used: Math.floor(new Date(timestamp).getTime() / 1000),
     };
     const deviceType = get(message, 'context.device.type');
```
Comment (mihir-4116) [inline src/v0/destinations/customerio/util.js:246]: what if we convert deviceType to string here itself?

**Review by mihir-4116 — CHANGES_REQUESTED** [review]

File: src/v0/destinations/customerio/util.js (hunk 3 above)
Comment (anantjain45823) [reply src/v0/destinations/customerio/util.js:246]: Do you mean this `deviceType = '123'` ?

File: src/v0/destinations/customerio/util.js (hunk 2 above)
Comment (anantjain45823) [reply src/v0/destinations/customerio/util.js:300]: No description is needed as the function name in itself is enough

**Review by mihir-4116 — APPROVED** [review]

**Review by koladilip — COMMENTED** [review]
Test cases?

**Comment by anantjain45823** [comment]
> Test cases?

Will first discuss these changes with @sanpj2292 once confirmed will add the test cases as well

File: src/v0/destinations/customerio/util.js [hunk 4]
```diff
@@ -236,13 +237,14 @@ const defaultResponseBuilder = (message, evName, userId, evType, destination, me
   // DEVICE registration
   const isDeviceRelatedEvent = isdeviceRelatedEventName(evName, destination);
   if (isDeviceRelatedEvent && userId && token) {
+    const timestamp = message.originalTimestamp;
```
Comment (krishna2020) [inline src/v0/destinations/customerio/util.js:240]: why a direct reference of originalTimestamp ? We use mapping config first to check on timestamp & then only fallback to originalTimestamp in all other destinations

Comment (anantjain45823) [reply src/v0/destinations/customerio/util.js]: Incorporated

**Review by mihir-4116 — APPROVED** [review]

File: test/__tests__/data/customerio_input.json [hunk 5]
```diff
@@ -2614,7 +2682,6 @@
       "request_ip": "[::1]",
       "rudderId": "0aef312c-0dc0-4a49-b613-4f33fb4e9b46",
       "sentAt": "2022-01-10T10:00:26.982Z",
```
Comment (krishna2020) [inline test/__tests__/data/customerio_input.json:2684]: shouldn't the right timestamp come here ? Instead how it's being omitted from the payload ?

**Comment by yashasvibajpai** [comment]
Do we plan to extend these to all or most used destinations too?

File: test/__tests__/data/customerio_input.json (hunk 5 above)
Comment (yashasvibajpai) [reply test/__tests__/data/customerio_input.json:2684]: We should revisit this, as the term `historical timestamp` seems very particular to customerIO here @anantjain45823

Comment (anantjain45823) [reply test/__tests__/data/customerio_input.json:2684]: > shouldn't the right timestamp come here ? Instead how it's being omitted from the payload ?

We have test cases for all type of timestamps following ISO format or timestamp as a number

Comment (anantjain45823) [reply test/__tests__/data/customerio_input.json:2684]: > We should revisit this, as the term `historical timestamp` seems very particular to customerIO here @anantjain45823

We are not changing the `historical timestamp` but the `created_at` field which excepts much old date as well

**Review by krishna2020 — APPROVED** [review]

**Comment by koladilip** [comment]
why is this still not merged?

**Review by sanpj2292 — APPROVED** [review]

**Pull request merged**